File: LeetCode/2368. Reachable Nodes With Restrictions/solution.cpp
class Solution {
public:
    int reachableNodes(int n, vector<vector<int>>& edges, vector<int>& r) {
        set<int>s(r.begin(),r.end());
        vector<vector<int>>to(n);
        for(auto i:edges)to[i[0]].push_back(i[1]),to[i[1]].push_back(i[0]);
        int ret=0;
        function<void(int,int)>dfs=[&](int p,int f){
            if(s.count(p))return;
            ++ret;
            for(int i:to[p])if(i!=f)dfs(i,p);
        };
        dfs(0,-1);
        return ret;
    }
};

File: LeetCode/2368. Reachable Nodes With Restrictions/solution.py
class Solution:
    def reachableNodes(self, n: int, edges: List[List[int]], restricted: List[int]) -> int:
        g = defaultdict(list)
        for a, b in edges:
            g[a].append(b)
            g[b].append(a)
        restricted = set(restricted)
        def dfs(cur, parent):
            if cur in restricted:
                return 0
            ans = 1
            for nxt in g[cur]:
                if nxt == parent:
                    continue
                ans += dfs(nxt, cur)
            return ans
        return dfs(0, 0)